compiler/lib/src/main/scala/codegen/JsonEncoder.scala

```diff
@@ -0,0 +1,698 @@
+package fpp.compiler
+
+import fpp.compiler.analysis._
+import fpp.compiler.ast._
+import fpp.compiler.codegen._
+import fpp.compiler.syntax._
+import fpp.compiler.transform._
+import fpp.compiler.util._
+import io.circe.syntax._
+import io.circe._
+import io.circe.generic.semiauto._
+import io.circe.generic.auto._
+import scala.util.parsing.input.Position
+
+object JsonEncoder {
+
+  /*
+    Defines implicit Encoder instances for various types in the Ast and PortInstance classes.
+    Each encoder returns a Unqualified Class Name Json.
+  */
+  implicit val binopEncoder: Encoder[Ast.Binop] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val unopEncoder: Encoder[Ast.Unop] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val componentKindEncoder: Encoder[Ast.ComponentKind] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val typeIntEncoder: Encoder[Ast.TypeInt] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val typeFloatEncoder: Encoder[Ast.TypeFloat] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val queueFullEncoder: Encoder[Ast.QueueFull] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val updateEncoder: Encoder[Ast.SpecTlmChannel.Update] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val limitKindEncoder: Encoder[Ast.SpecTlmChannel.LimitKind] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val generalKindEncoder: Encoder[Ast.SpecPortInstance.GeneralKind] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val specialInputKindEncoder: Encoder[Ast.SpecPortInstance.SpecialInputKind] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val specialKindEncoder: Encoder[Ast.SpecPortInstance.SpecialKind] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val visibilityEncoder: Encoder[Ast.Visibility] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val formalParamKindEncoder: Encoder[Ast.FormalParam.Kind] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val generalPortInstanceKindEncoder: Encoder[PortInstance.General.Kind] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val patternEncoder: Encoder[Ast.SpecConnectionGraph.Pattern.Kind] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+  implicit val severityEncoder: Encoder[Ast.SpecEvent.Severity] =
+    Encoder.encodeString.contramap(getUnqualifiedClassName(_))
+
+  implicit val typeEncoder: Encoder[Type] =
```

Comment:
I don't think we want to use toString here. We should preserve the representation of these objects as data structures.

Reply:
I also think these encoders belong with the analysis encoders below. Also, if possible we should use object {...} blocks to group the Ast encoders separately from the Analysis encoders.

Reply:
Can we move the Ast, Analysis, and LocMap encoders each into its own file? Then JsonEncoder would become a subdirectory, like CppWriter, etc.